**Dealing with duplicated values**

In [90]:
import pandas as pd

df = pd.read_csv('/content/sample_data/part1.csv', index_col='index')
df_copy = df.copy()

In [91]:
df_copy.head()

order id      date                        status fulfilment  \
index                                                                           
0      405-8078784-5731545  04-30-22                     Cancelled   Merchant   
1      171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   Merchant   
2      404-0687676-7273146  04-30-22                       Shipped     Amazon   
3      403-9615377-8133951  04-30-22                     Cancelled   Merchant   
4      407-1069790-7240320  04-30-22                       Shipped     Amazon   

      sales channel  ship-service-level    style              sku  \
index                                                               
0          Amazon.in           Standard   SET389   SET389-KR-NP-S   
1          Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL   
2          Amazon.in          Expedited  JNE3371    JNE3371-KR-XL   
3          Amazon.in           Standard    J0341       J0341-DR-L   
4          Amazon.in          Expedited  JNE3671  JNE3671-TU-XXXL   

            category size        asin courier status  qty currency  amount  \
index                                                                        
0                Set    S  B09KXVBD7Z        Unknown    0      INR  647.62   
1              kurta  3XL  B09K3WFS32        Shipped    1      INR  406.00   
2              kurta   XL  B07WV4JV4D        Shipped    1      INR  329.00   
3      Western Dress    L  B099NRCT7B        Unknown    0      INR  753.33   
4                Top  3XL  B098714BZP        Shipped    1      INR  574.00   

         ship-city   ship-state  ship-postal-code ship-country    b2b  
index                                                                  
0           MUMBAI  MAHARASHTRA            400081           IN  False  
1        BENGALURU    KARNATAKA            560085           IN  False  
2      NAVI MUMBAI  MAHARASHTRA            410210           IN   True  
3       PUDUCHERRY   PUDUCHERRY            605008           IN  False  
4          CHENNAI   TAMIL NADU            600073           IN  False

Let's check how many unique values are in DataFrame

In [92]:
df_copy.nunique()

order id              120378
date                      91
status                    13
fulfilment                 2
sales channel              2
ship-service-level         2
style                   1377
sku                     7195
category                   9
size                      11
asin                    7190
courier status             4
qty                       10
currency                   1
amount                  1411
ship-city               8956
ship-state                70
ship-postal-code        9460
ship-country               1
b2b                        2
dtype: int64

The first step was to determine the **grain** of the dataset — that is, what a single row represents.

In [93]:
is_order_grain = df_copy['order id'].nunique() / len(df_copy) * 100
is_order_grain

93.3343671254119

The result was 93.33%, meaning that approximately 6.67% of rows share the same order id.

In [94]:
df_copy['order id'].value_counts().sort_values(ascending=False)

order id
403-4984515-8861958    12
171-5057375-2831560    12
403-0173977-3041148    11
404-9932919-6662730    11
408-3317403-1729937    10
                       ..
406-1589049-9362704     1
403-0275824-1655546     1
404-4885091-8361169     1
408-0886405-2705923     1
407-3807112-3689150     1
Name: count, Length: 120378, dtype: int64

As we can see, we have duplicated order_id's - one order may have multiple rows. <br>Grain != Order.<br> Let's analyze what happens during filtering by one order_id.

In [102]:
df_copy[df_copy['order id'] == '403-4984515-8861958']

order id      date   status fulfilment sales channel   \
31736  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31737  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31738  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31739  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31740  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31741  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31742  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31743  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31744  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31745  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31746  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   
31747  403-4984515-8861958  04-11-22  Shipped     Amazon      Amazon.in   

      ship-service-level    style                sku       category size  \
31736          Expedited  JNE3792    JNE3792-KR-XXXL          kurta  3XL   
31737          Expedited  JNE3764    JNE3764-KR-XXXL          kurta  3XL   
31738          Expedited  JNE3503    JNE3503-KR-XXXL          kurta  3XL   
31739          Expedited    J0308      J0308-DR-XXXL  Western Dress  3XL   
31740          Expedited  JNE3702    JNE3702-KR-XXXL          kurta  3XL   
31741          Expedited  JNE3741    JNE3741-KR-XXXL          kurta  3XL   
31742          Expedited  JNE3634    JNE3634-KR-XXXL          kurta  3XL   
31743          Expedited   SET433  SET433-KR-NP-XXXL            Set  3XL   
31744          Expedited  JNE3636    JNE3636-KR-XXXL          kurta  3XL   
31745          Expedited  JNE3633    JNE3633-KR-XXXL          kurta  3XL   
31746          Expedited  JNE3813    JNE3813-KR-XXXL          kurta  3XL   
31747          Expedited  JNE3638    JNE3638-KR-XXXL          kurta  3XL   

             asin courier status  qty currency  amount ship-city   ship-state  \
31736  B09KXNWRSB        Shipped    1      INR   432.0   SOLAPUR  MAHARASHTRA   
31737  B09K3SSDSZ        Shipped    1      INR   487.0   SOLAPUR  MAHARASHTRA   
31738  B08RP2BV5Q        Shipped    1      INR   318.0   SOLAPUR  MAHARASHTRA   
31739  B099NTV5ZV        Shipped    1      INR   665.0   SOLAPUR  MAHARASHTRA   
31740  B093ZRBGR9        Shipped    1      INR   342.0   SOLAPUR  MAHARASHTRA   
31741  B099NL77RF        Shipped    1      INR   432.0   SOLAPUR  MAHARASHTRA   
31742  B097ZZTMZD        Shipped    1      INR   511.0   SOLAPUR  MAHARASHTRA   
31743  B09TH8T4GM        Shipped    1      INR   666.0   SOLAPUR  MAHARASHTRA   
31744  B09812FYH9        Shipped    1      INR   459.0   SOLAPUR  MAHARASHTRA   
31745  B094FQ32VX        Shipped    1      INR   459.0   SOLAPUR  MAHARASHTRA   
31746  B09SDXYXVK        Shipped    1      INR   635.0   SOLAPUR  MAHARASHTRA   
31747  B09812MM2Q        Shipped    1      INR   517.0   SOLAPUR  MAHARASHTRA   

       ship-postal-code ship-country    b2b  
31736            413002           IN  False  
31737            413002           IN  False  
31738            413002           IN  False  
31739            413002           IN  False  
31740            413002           IN  False  
31741            413002           IN  False  
31742            413002           IN  False  
31743            413002           IN  False  
31744            413002           IN  False  
31745            413002           IN  False  
31746            413002           IN  False  
31747            413002           IN  False

Order id may be the same but we have other different values. Ex: sku. One order - many different products.

In [96]:
df_copy.duplicated().sum() # 6 rows are full duplicates of other rows

np.int64(6)

In [97]:
df_copy[df_copy.duplicated(subset=['order id'], keep=False)].sort_values(by='order id')

order id      date                        status fulfilment  \
index                                                                           
78939  171-0012693-2893106  05-07-22                     Cancelled   Merchant   
78938  171-0012693-2893106  05-07-22                     Cancelled   Merchant   
91476  171-0014231-7760344  06-29-22                       Pending     Amazon   
91475  171-0014231-7760344  06-29-22                       Pending     Amazon   
24515  171-0015435-1981904  04-16-22                       Shipped     Amazon   
...                    ...       ...                           ...        ...   
74748  408-9929975-2118741  05-11-22  Shipped - Delivered to Buyer   Merchant   
36052  408-9951521-7399501  04-09-22                       Shipped     Amazon   
36053  408-9951521-7399501  04-09-22                       Shipped     Amazon   
61445  408-9954515-8573127  05-22-22  Shipped - Delivered to Buyer   Merchant   
61444  408-9954515-8573127  05-22-22  Shipped - Delivered to Buyer   Merchant   

      sales channel  ship-service-level    style                sku  \
index                                                                 
78939      Amazon.in           Standard    NW015      NW015-TP-PJ-S   
78938      Amazon.in           Standard    NW030      NW030-TP-PJ-S   
91476      Amazon.in          Expedited    J0224       J0224-SKD-XS   
91475      Amazon.in          Expedited   SET342  SET342-KR-NP-N-XS   
24515      Amazon.in          Expedited  JNE3611       JNE3611-KR-L   
...              ...                ...      ...                ...   
74748      Amazon.in           Standard   SET408  SET408-KR-NP-XXXL   
36052      Amazon.in          Expedited    J0338         J0338-DR-M   
36053      Amazon.in          Expedited    J0335         J0335-DR-M   
61445      Amazon.in           Standard   SET356   SET356-KR-NP-XXL   
61444      Amazon.in           Standard   SET356     SET356-KR-NP-M   

            category size        asin courier status  qty currency   amount  \
index                                                                         
78939            Set    S  B0922SCDFN        Unknown    0      INR   477.14   
78938            Set    S  B09G2RFP14        Unknown    0      INR   538.10   
91476            Set   XS  B09RKFBZZD      Unshipped    1      INR  1364.00   
91475            Set   XS  B09Q1943PQ        Shipped    1      INR   850.00   
24515          kurta    L  B08XVQMSJV        Shipped    1      INR   459.00   
...              ...  ...         ...            ...  ...      ...      ...   
74748            Set  3XL  B09RKF7PZB        Shipped    1      INR   499.00   
36052  Western Dress    M  B0982ZPB81        Shipped    1      INR   744.00   
36053  Western Dress    M  B09831XC46        Shipped    1      INR   807.00   
61445            Set  XXL  B09QJ3RNNY        Shipped    1      INR   999.00   
61444            Set    M  B09QJ4D21Y        Shipped    1      INR   999.00   

              ship-city          ship-state  ship-postal-code ship-country  \
index                                                                        
78939    SRI GANGANAGAR           RAJASTHAN            335001           IN   
78938    SRI GANGANAGAR           RAJASTHAN            335001           IN   
91476         AHMEDABAD             Gujarat            380015           IN   
91475         AHMEDABAD             Gujarat            380015           IN   
24515  Pimpri Chinchwad         MAHARASHTRA            412101           IN   
...                 ...                 ...               ...          ...   
74748        port blair  ANDAMAN & NICOBAR             744101           IN   
36052      SECUNDERABAD           TELANGANA            500017           IN   
36053      SECUNDERABAD           TELANGANA            500017           IN   
61445            SHIMLA    HIMACHAL PRADESH            171005           IN   
61444            SHIMLA    HIMACHAL PRADESH            171005           IN  

Filtering duplicates based on order_id and sku.

In [98]:
df_copy[df_copy.duplicated(subset=['order id', 'sku'], keep=False)]

order id      date     status fulfilment sales channel   \
index                                                                       
30660  406-0372545-6086735  04-12-22  Cancelled     Amazon      Amazon.in   
30661  406-0372545-6086735  04-12-22  Cancelled     Amazon      Amazon.in   
41291  408-0373839-4433120  04-05-22  Cancelled     Amazon      Amazon.in   
41292  408-0373839-4433120  04-05-22  Cancelled     Amazon      Amazon.in   
52330  407-8364731-6449117  05-29-22    Shipped     Amazon      Amazon.in   
52333  407-8364731-6449117  05-29-22    Shipped     Amazon      Amazon.in   
79844  171-9628368-5329958  05-07-22  Cancelled     Amazon      Amazon.in   
79845  171-9628368-5329958  05-07-22  Cancelled     Amazon      Amazon.in   
85790  171-3249942-2207542  05-03-22    Shipped     Amazon      Amazon.in   
85791  171-3249942-2207542  05-03-22    Shipped     Amazon      Amazon.in   
86418  405-8669298-3850736  05-03-22    Shipped     Amazon      Amazon.in   
86419  405-8669298-3850736  05-03-22    Shipped     Amazon      Amazon.in   
98954  407-4853873-4978725  06-22-22    Shipped     Amazon      Amazon.in   
98955  407-4853873-4978725  06-22-22    Shipped     Amazon      Amazon.in   

      ship-service-level    style              sku category size        asin  \
index                                                                          
30660          Expedited   SET197   SET197-KR-NP-L      Set    L  B08B3YPD63   
30661          Expedited   SET197   SET197-KR-NP-L      Set    L  B08B3YPD63   
41291          Expedited  JNE3501     JNE3501-KR-M    kurta    M  B08MYVCDB4   
41292          Expedited  JNE3501     JNE3501-KR-M    kurta    M  B08MYVCDB4   
52330          Expedited  JNE3769     JNE3769-KR-L    kurta    L  B09K3T675B   
52333          Expedited  JNE3769     JNE3769-KR-L    kurta    L  B09K3T675B   
79844          Expedited    J0329       J0329-KR-L    kurta    L  B09KXRB7JV   
79845          Expedited    J0329       J0329-KR-L    kurta    L  B09KXRB7JV   
85790          Expedited   SET323  SET323-KR-NP-XL      Set   XL  B09NDKKM7P   
85791          Expedited   SET323  SET323-KR-NP-XL      Set   XL  B09NDKKM7P   
86418          Expedited  MEN5025  MEN5025-KR-XXXL    kurta  3XL  B08YYQS8FH   
86419          Expedited  MEN5025  MEN5025-KR-XXXL    kurta  3XL  B08YYQS8FH   
98954          Expedited    J0230      J0230-SKD-M      Set    M  B08XNJG8B1   
98955          Expedited    J0230      J0230-SKD-M      Set    M  B08XNJG8B1   

      courier status  qty currency   amount  ship-city     ship-state  \
index                                                                   
30660      Cancelled    0      INR   648.56   Siliguri    WEST BENGAL   
30661      Cancelled    0      INR   648.56   Siliguri    WEST BENGAL   
41291      Cancelled    0      INR   648.56  BENGALURU      KARNATAKA   
41292      Cancelled    0      INR   648.56  BENGALURU      KARNATAKA   
52330      Cancelled    0      INR   648.56      Theni     TAMIL NADU   
52333        Shipped    1      INR   487.00      Theni     TAMIL NADU   
79844      Cancelled    0      INR   648.56  ERNAKULAM         KERALA   
79845      Cancelled    0      INR   648.56  ERNAKULAM         KERALA   
85790        Shipped    1      INR   939.00       PUNE    MAHARASHTRA   
85791        Shipped    1      INR   939.00       PUNE    MAHARASHTRA   
86418        Shipped    1      INR   533.00  GHAZIABAD  UTTAR PRADESH   
86419        Shipped    1      INR   533.00  GHAZIABAD  UTTAR PRADESH   
98954        Shipped    1      INR  1163.00   Zirakpur         Punjab   
98955        Shipped    1      INR  1163.00   Zirakpur         Punjab   

       ship-postal-code ship-country    b2b  
index                                        
30660            734008           IN  False  
30661            734008           IN  False  
41291            560072           IN  False  
41292            560072           IN  False  
52330            625579           IN  False  
52333        

There we can see that we have 6 exact duplicates that we can get removed but also a business case: <br>
52330 / 52333 <br>
order id = 407-8364731-6449117 <br>
sku = JNE3769-KR-L <br>
that may be a refund or shipping problem case. This row will be further analyzed.

In [99]:
df_copy = df_copy.drop_duplicates(keep='first').reset_index(drop=True)

In [100]:
#check
df_copy.duplicated(subset=['order id','sku']).sum()

np.int64(1)

In [101]:
df_copy.to_csv('sample_data/part2.csv')

Summary:
**grain**: One row represents one product (SKU) within one order.
<br>
Logical key = (order id, sku) <br>
So, grain is reprezented by one product within order with it's status. Repeated order id values are expected due to multi-product orders.
Duplicates are dropped.